### PR TITLE
Async improvements + Async Inn Sequence

### PR DESCRIPTION
**Pull request by fmatthew5876**
~~Depends on #1742~~

The PR has 2 goals:
* Refactor the async framework:
    * Don't use globals
    * Kill `transition` game temp vars
    * Support parameters for async ops
* Change Inn Fade Out, BGM, Fade In sequence to be asynchronous

Test Cases of #1689 
- [x] Test 1
- [ ] Test 2 - Requires message fixes - to be done in a later PR after message refactor commits get in.
- [x] Tests 3 - 5
- [ ] Test 6 - not sure we even want to emulate this..?

### CODE REVIEW

**Comment by fmatthew5876** [comment]
I retested #1819 and Ghabry's async test case from #1784 

I rebased a local copy of this onto #1881 and retested both #1261 and #1840 

This is as much as I can do for the Inn until some of the message work is done.

This PR is done and ready for review.

**Comment by Ghabry** [comment]
RFC: Is there any way to combine this with the Async code we have for emscripten or do you consider them too different?

**Comment by fmatthew5876** [comment]
> RFC: Is there any way to combine this with the Async code we have for emscripten or do you consider them too different?

Which async code are you referring to specifically?

This one is a sort of poor mans way of turning our game loop into a resumable co-routine.

**Comment by Ghabry** [comment]
I mean the Async code for downloading files that emscripten calls (async_handler.cpp)

**Comment by fmatthew5876** [comment]
They already kind of work together since file IO works the same as animating transitions.

Do we have any case where we would need to suspend and resume the game loop at a specific point for a file load? I could add an `AsyncOp::eFileLoad` to do that.

It would likely cause the game to stutter though, as everything would have to freeze in the middle of the frame until the IO finishes.

As far as I'm aware, most things in the game logic don't depend on file IO to be done and are fine waiting until the end of the frame.

**Comment by Ghabry** [comment]
No, we already suspend via "ImportantFilePending" when the game loop returns and for non-important files which don't suspend there is code which delays rendering operations on these files (e.g. when Width/height are necessary), but update code still runs as expected. 
Just thought that this can be maybe simplified by merging it with the new AsyncOp code, but not necessary.

File: src/scene_map.cpp [hunk 1]
```diff
@@ -392,6 +396,58 @@ void Scene_Map::OnAsyncSuspend(F&& f, AsyncOp aop, bool is_preupdate) {
...
+	}
+
 	AsyncNext(std::forward<F>(f));
 }
 
+void Scene_Map::StartInn() {
+	const RPG::Music& bgm_inn = Game_System::GetSystemBGM(Game_System::BGM_Inn);
+	if (bgm_inn.name.empty() ||
+		bgm_inn.name == "(OFF)" ||
+		bgm_inn.name == "(Brak)") {
```
Comment (Ghabry) [inline src/scene_map.cpp]: I wanted to complain now that this is wrong but this is even in the legacy code.

The new way to detect if a string is (OFF) or similiar is "isStopFilename" in Game_System but the function is static and not exported yet.

File: src/scene_map.cpp [hunk 2]
```diff
@@ -392,6 +396,58 @@ void Scene_Map::OnAsyncSuspend(F&& f, AsyncOp aop, bool is_preupdate) {
...
+		FinishInn();
+		return;
+	}
+
+	Game_System::BgmStop();
```
Comment (Ghabry) [inline src/scene_map.cpp]: It's not necessary to call BgmStop before BgmPlay

Comment (fmatthew5876) [reply src/scene_map.cpp]: Fixed

File: src/scene_map.cpp (hunk 1 above)
Comment (fmatthew5876) [reply src/scene_map.cpp]: Fixed

**Comment by fmatthew5876** [comment]
I pushed another change which creates a `Scene_Inn` for the inn logic. Now the inn logic is separated from `Scene_Map`. This change allows us to use push and pop scenes within async operations.

This will enable #1891 to be implemented using this same machinery.

**Comment by fmatthew5876** [comment]
Redid the inn test cases.

This is ready again

**Comment by fmatthew5876** [comment]
I took out the async `Scene_Inn`. I'm still on the fence about whether it's good idea to support asynchronous scene calling.

**Pull request merged**